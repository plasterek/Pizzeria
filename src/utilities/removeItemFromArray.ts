export const removeItemFromArray = <T>(array: T[], item: T): void => {
  const indexOfItem = array.findIndex((items) => items === item);
  array.splice(indexOfItem, 1);
};
