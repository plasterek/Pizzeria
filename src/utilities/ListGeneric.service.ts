import {
  IGenericObject,
  TGenericProperty,
  IGenericListObject,
} from "./models/Generics.model";
import { removeItemFromArray } from "./removeItemFromArray";

export class List<T extends IGenericObject, K extends TGenericProperty> {
  private list: IGenericListObject<T, K>[] = [];

  public add(object: T, property: K): void {
    this.list.push({ object, property });
  }
  public delete(object: T): void {
    const itemExist: IGenericListObject<T, K> | undefined = this.getObjectById(
      object.id
    );
    if (!itemExist) {
      throw new Error();
    }
    removeItemFromArray(this.list, itemExist);
  }
  getList(): IGenericListObject<T, K>[] {
    return this.list;
  }
  public getObjectById(objectId: string): IGenericListObject<T, K> | undefined {
    const objectExist: IGenericListObject<T, K> | undefined = this.list.find(
      (element) => element.object.id === objectId
    );
    return objectExist;
  }
}
