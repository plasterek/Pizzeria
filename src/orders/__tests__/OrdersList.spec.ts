import { IOrder } from "../models/Order.model";
import { OrdersList } from "../OrdersList.service";
import * as removeItemFromArray from "../../utilities/removeItemFromArray";

describe("Orders list class", () => {
  //given
  const waiterId: string = "1";
  const pizza: Map<string, number> = new Map().set("pizza", 1);
  const order: IOrder = { id: "orderId", pizzaWithQuantity: pizza, tableId: "tableId", cookId: "cookId" };
  let ordersList = new OrdersList();
  beforeEach(() => {
    ordersList = new OrdersList();
    jest.clearAllMocks();
  });
  describe("When trying to add an order and order already exist on list", () => {
    it("It should throw an exception", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      //then
      expect(() => ordersList.makeOrder(order, waiterId)).toThrow();
    });
  });
  describe("When trying to add an order and everything went well", () => {
    it("Order should appear on order list", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      //then
      expect(ordersList).toMatchSnapshot();
    });
    it("getObjectById method should be called", () => {
      //given
      const getObject = jest.spyOn(ordersList, "getObjectById");
      //when
      ordersList.makeOrder(order, waiterId);
      //then
      expect(getObject).toBeCalled();
    });
    it("add method should be called", () => {
      //given
      const add = jest.spyOn(ordersList, "add");
      //when
      ordersList.makeOrder(order, waiterId);
      //then
      expect(add).toBeCalled();
    });
  });
  describe("When trying to get list of pizzas waiting in que", () => {
    it("It should return array with IDs or empty array", () => {
      //given
      const getPizzas: string[] = ordersList.getWaitingPizzas();
      //then
      expect(getPizzas).toMatchSnapshot();
    });
  });
  describe("When trying to add order to waitlist and order does not exist in orders list", () => {
    it("It should throw an exception", () => {
      //then
      expect(() => ordersList.addToWaitList(order.id)).toThrow();
    });
  });
  describe("When trying to add order to waitlist and order was already added to waitlist", () => {
    it("It should throw an exception", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToWaitList(order.id);
      //then
      expect(() => ordersList.addToWaitList(order.id)).toThrow();
    });
  });
  describe("When trying to add order to waitlist and everything went well", () => {
    it("order id should be added to waitlist", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToWaitList(order.id);
      //then
      expect(ordersList.getWaitingPizzas().length).toBeGreaterThan(0);
    });
    it("getObjectById method should be called", () => {
      //given
      const getObject = jest.spyOn(ordersList, "getObjectById");
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToWaitList(order.id);
      //then
      expect(getObject).toBeCalled();
    });
  });
  describe("When trying to add order to realization and order does not exist on orders list", () => {
    it("It should throw an exception", () => {
      //then
      expect(() => ordersList.addToRealisation(order.id)).toThrow();
    });
  });
  describe("When trying to add order to realization and order was already added to realization", () => {
    it("It should throw an exception", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToRealisation(order.id);
      //then
      expect(() => ordersList.addToRealisation(order.id)).toThrow();
    });
  });
  describe("When trying to add order to realization and order was already added to waitlist", () => {
    it("Order id should be removed from waitlist", () => {
      //given
      const waitList: string[] = ordersList.getWaitingPizzas();
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToWaitList(order.id);
      ordersList.addToRealisation(order.id);
      //then
      expect(waitList).toMatchSnapshot();
    });
    it("removeItemFromArray function should be called", () => {
      //given
      const remove = jest.spyOn(removeItemFromArray, "removeItemFromArray");
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToWaitList(order.id);
      ordersList.addToRealisation(order.id);
      //then
      expect(remove).toBeCalled();
    });
  });
  describe("When trying to add order to realization and everything went well", () => {
    it("Order id should be added to orders in realization list", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToRealisation(order.id);
      //then
      expect(ordersList).toMatchSnapshot();
    });
  });
  describe("When trying to finish order and order does not exist on order list", () => {
    it("It should throw an exception", () => {
      //then
      expect(() => ordersList.finishOrder(order.id)).toThrow();
    });
  });
  describe("When trying to finish order and order was already finished", () => {
    it("It should throw an exception", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToRealisation(order.id);
      ordersList.finishOrder(order.id);
      //then
      expect(() => ordersList.finishOrder(order.id)).toThrow();
    });
  });
  describe("When trying to finish order and everything went well", () => {
    it("removeItemFromArray function should be called", () => {
      //given
      const remove = jest.spyOn(removeItemFromArray, "removeItemFromArray");
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToRealisation(order.id);
      ordersList.finishOrder(order.id);
      //then
      expect(remove).toBeCalled();
    });
    it("Given order id should be removed from realization list and appear in finished list", () => {
      //when
      ordersList.makeOrder(order, waiterId);
      ordersList.addToRealisation(order.id);
      ordersList.finishOrder(order.id);
      //then
      expect(ordersList).toMatchSnapshot();
    });
  });
});
