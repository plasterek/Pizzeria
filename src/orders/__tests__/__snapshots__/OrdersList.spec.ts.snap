// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders list class When trying to add an order and everything went well Order should appear on order list 1`] = `
OrdersList {
  "finished": [],
  "inRealization": [],
  "list": [
    {
      "object": {
        "cookId": "cookId",
        "id": "orderId",
        "pizzaWithQuantity": Map {
          "pizza" => 1,
        },
        "tableId": "tableId",
      },
      "property": "1",
    },
  ],
  "waiting": [],
}
`;

exports[`Orders list class When trying to add order to realization and everything went well Order id should be added to orders in realization list 1`] = `
OrdersList {
  "finished": [],
  "inRealization": [
    "orderId",
  ],
  "list": [
    {
      "object": {
        "cookId": "cookId",
        "id": "orderId",
        "pizzaWithQuantity": Map {
          "pizza" => 1,
        },
        "tableId": "tableId",
      },
      "property": "1",
    },
  ],
  "waiting": [],
}
`;

exports[`Orders list class When trying to add order to realization and order was already added to waitlist Order id should be removed from waitlist 1`] = `[]`;

exports[`Orders list class When trying to finish order and everything went well Given order id should be removed from realization list and appear in finished list 1`] = `
OrdersList {
  "finished": [
    "orderId",
  ],
  "inRealization": [],
  "list": [
    {
      "object": {
        "cookId": "cookId",
        "id": "orderId",
        "pizzaWithQuantity": Map {
          "pizza" => 1,
        },
        "tableId": "tableId",
      },
      "property": "1",
    },
  ],
  "waiting": [],
}
`;

exports[`Orders list class When trying to get list of pizzas waiting in que It should return array with IDs or empty array 1`] = `[]`;
